RavenDB-22624 - Add RavenQuery.Id method to allow explicit identity projection

golan22888 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22624/Add-RavenQuery.Id-method-to-allow-explicit-identity-projection

### Additional description

Add `RavenQuery.Id` method to allow explicit identity projection.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

Commit:
RavenDB-22624 - Added RavenQuery.Id

grisha-kotler commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -2566,11 +2566,18 @@ private static bool CanConvert(Expression expression, DocumentConventions conven
+                if (expression is MethodCallExpression { Method.Name: nameof(RavenQuery.Id) } mce && mce.Method.DeclaringType == typeof(RavenQuery) )
+                {
+                    if (mce.Arguments.Count != 1)
+                        throw new InvalidQueryException($"{nameof(RavenQuery.Id)} can only get one argument");
```
```suggestion
                        throw new InvalidOperationException($"{nameof(RavenQuery.Id)} can only get one argument");
```

golan22888 replied:
Done.

Commit:
RavenDB-22624 - changed the exception to InvalidOperationExeption.

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -2806,6 +2812,12 @@ private void SelectNewExpression(NewExpression newExpression, LambdaExpression l
+
+                    if (LinqPathProvider.IsIdCall(mce))
+                    {
+                        HandleSelectId( GetSelectPath(newExpression.Members[index]));
```
formatting

golan22888 replied:
Done.

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -2896,6 +2908,13 @@ private void HandleSelectTimeSeries(MethodCallExpression callExpression, string
 
+        private void HandleSelectId( string alias = null)
+        {
+            string path = $"{Constants.Documents.Indexing.Fields.DocumentIdFieldName}";
```
assuming that this can be without `$""` just: `string path = Constants.Documents.Indexing.Fields.DocumentIdFieldName` ?

golan22888 replied:
Done.

ppekrol commented on test/FastTests/Client/RavenQueryTests.cs:
```diff
@@ -0,0 +1,166 @@
+
+                        Assert.Equal(
+                            "from index 'Employees/ByFirstName' as a select { FullName " +
+                            ": a.FirstName+\" \"+a.LastName, EmployeeId : id(a) }",
```
can we put it in 1 line without concatenation?

golan22888 replied:
Done.

ppekrol commented on test/FastTests/Client/RavenQueryTests.cs:
```diff
@@ -0,0 +1,166 @@
+                        var returnedEmployee32 = returnedEmployees3.Last();
+
+                        Assert.Equal(
+                            "from index 'Employees/ByFirstName' as a select { FullName : " +
```
can we put it in 1 line without concatenation?

golan22888 replied:
Done.

ppekrol commented on test/FastTests/Client/RavenQueryTests.cs:
```diff
@@ -0,0 +1,166 @@
+                        var returnedEmployee42 = returnedEmployees4.Last();
+
+                        Assert.Equal(
+                            "from 'NewEmployees' as a select { EmployeeId : id(a), FullName : " +
```
can we put it in 1 line without concatenation?

golan22888 replied:
Done.

ppekrol commented on test/FastTests/Client/RavenQueryTests.cs:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        [RavenTheory(RavenTestCategory.ClientApi)]
+        [RavenData(DatabaseMode = RavenDatabaseMode.All)]
```
no need to test in in Sharding I think?

golan22888 replied:
Done.

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -2896,6 +2908,13 @@ private void HandleSelectTimeSeries(MethodCallExpression callExpression, string
             AddToFieldsToFetch(path, alias);
         }
 
+        private void HandleSelectId( string alias = null)
```
formatting :)

golan22888 replied:
Done.

Lwiel reviewed (commented):
Can we add a test for the [original case](https://issues.hibernatingrhinos.com/issue/RavenDB-22556/JS-projection-query-uses-x.Id-on-6.0-while-it-used-idx-on-5.4)?

Commit:
RavenDB-22624
formatted the code and add the original case of issue 22556.
added a test with chaining select too.

Commit:
RavenDB-22624: changed back that a.Id translated to a.id instead of id(a) and RavenQuery.Id(a) translate to id(a).
